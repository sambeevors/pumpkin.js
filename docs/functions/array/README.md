---
sidebarDepth: 0
next: /functions/dom/
prev: /functions/string/
---

# Array functions

The following functions are array specific.

## first

Get the first item(s) in an array

### Parameters

- `array` (type: Array) - The array to select from
- `count` (type: Number) - The number of items to retrieve (optional)
- `reverse` (type: Boolean) - Whether to reverse the returned array (optional)

### Returns

- (type: Any) - Either first item or an array of first x items

### Example

```js
const arr = [1, 2, 3, 4, 5]

$.first(arr) // [1]
$.first(arr, 2) // [1, 2]
$.first(arr, 2, true) // [2, 1]

// Also works with node arrays generated by the qsa function!
$foos = $.qsa('.foo')
$firstFoo = $.first($foos)
```

## last

Get the last item(s) in an array

### Parameters

- `array` (type: Array) - The array to select from
- `count` (type: Number) - The number of items to retrieve (optional)
- `reverse` (type: Boolean) - Whether to reverse the returned array (optional)

### Returns

- (type: Any) - Either last item or an array of last x items

### Example

```js
const arr = [1, 2, 3, 4, 5]

$.last(arr) // [5]
$.last(arr, 2) // [4, 5]
$.last(arr, 2, true) // [5, 4]

// Also works with node arrays generated by the qsa function!
$foos = $.qsa('.foo')
$lastFoo = $.last($foos)
```

## strip

Remove value(s) from an array

### Parameters

- `array` (type: Array) - The array to select from
- `values` (type: Any) - The values to remove

### Returns

- (type: Array) - The new array

### Example

```js
const arr = [1, 2, 3, 4, 5]

$.strip(arr, 2) // [1, 3, 4, 5]
$.strip(arr, 2, 5) // [1, 3, 4]
```

## types

Get the type of each value in an array

### Parameters

- `array` (type: Array) - The array to use

### Returns

- (type: Array) - An array of item types

### Example

```js
const arr = [1, true, 'foo', {}, Symbol(), () => true]

$.types(arr) // ['number', 'boolean', 'string', 'object', 'symbol', 'function']

/**
 * Can also be used to test types of non-arrays,
 * but not reccomended. Use typeof instead.
 */
$.types('foo') // 'string'
```
