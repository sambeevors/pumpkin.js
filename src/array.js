/**
 * Get the first item(s) in an array
 * @param  {Array} array     The array to select from
 * @param  {Number} count    The number of items to retrieve [optional]
 * @param  {Boolean} reverse Whether to reverse the returned array [optional]
 * @return {Any}             Either first item or an array of first x items
 */
export const first = (array, count, reverse) => {
  if (!count) return array[0]

  let newArray = []
  for (let i = 0; i < count; i++) newArray.push(array[i])
  return reverse ? newArray.reverse() : newArray
}

/**
 * Get the last item(s) in an array
 * @param  {Array} array     The array to select from
 * @param  {Number} count    The number of items to retrieve [optional]
 * @param  {Boolean} reverse Whether to reverse the returned array [optional]
 * @return {Any}             Either last item or an array of last x items
 */
export const last = (array, count, reverse) => {
  if (!count) return array[array.length - 1]

  let newArray = []
  for (let i = 1; i <= count; i++) newArray.push(array[array.length - i])
  return reverse ? newArray : newArray.reverse()
}

/**
 * Remove value(s) from an array
 * @param  {Array} array The array to select from
 * @param  {Any} values  The values to remove
 * @return {Array}       The new array
 */
export const strip = (array, ...values) => {
  values.forEach(value => {
    array = array.filter(item => item !== value)
  })
  return array
}
