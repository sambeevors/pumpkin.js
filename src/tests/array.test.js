import * as $ from '../array'
import DOM from './helpers/fake-dom'

let $toDoList
let $toDoListItems
let $toDoListItem

const dom = new DOM(
  `<ul id="to-do-list">
    <li>Go shopping</li>
    <li>Do laundry</li>
    <li>Raid fridge!</li>
  </ul>`,
  () => {
    $toDoList = document.querySelector('#to-do-list')
    $toDoListItems = [...$toDoList.querySelectorAll('li')]
    $toDoListItem = $toDoList.querySelector('li')
  }
)

test('First Item(s) In Array', () => {
  // Basic test
  expect($.first([1, 2, 3, 4])).toBe(1)
  // Test with count parameter
  expect($.first([1, 2, 3, 4], 2)).toContain(1, 2)
  // Test with reverse parameter
  expect($.first([1, 2, 3, 4], 2, true)).toContain(2, 1)
  // Test with DOM nodes
  expect($.first($toDoListItems)).toMatchSnapshot()
  // Test with invalid parameter
  expect($.first([1, 2], 3, true)[0]).toBeUndefined()
})

test('Last Item(s) In Array', () => {
  // Basic test
  expect($.last([1, 2, 3, 4])).toBe(4)
  // Test with count parameter
  expect($.last([1, 2, 3, 4], 2)).toContain(3, 4)
  // Test with reverse parameter
  expect($.last([1, 2, 3, 4], 2, true)).toContain(4, 3)
  // Test with DOM nodes
  expect($.last($toDoListItems)).toMatchSnapshot()
  // Test with invalid parameter
  expect($.last([1, 2], 3)[0]).toBeUndefined()
})

test('Strip Items(s) From Array', () => {
  // Basic test
  expect($.strip([1, 2, 3, 4], 1)).toContain(2, 3, 4)
  // Test with extra parameters
  expect($.strip([1, 2, 3, 4], 1, 3)).toContain(2, 4)
  // Test with non existant value
  expect($.strip([1, 2, 3, 4], 5)).toContain(1, 2, 3, 4)
  // Test with DOM nodes
  expect($.strip($toDoListItems, $toDoListItem)).toMatchSnapshot()
})

test('Get Types Of Elements In Array', () => {
  // Basic test
  let undefined
  expect(
    $.types([1, true, 'foo', {}, Symbol(), () => true, undefined])
  ).toEqual([
    'number',
    'boolean',
    'string',
    'object',
    'symbol',
    'function',
    'undefined'
  ])
  // Test with non-array
  expect($.types('foo')).toBe('string')
  // Test with empty array
  expect($.types([])).toEqual([])
  // Test with empty object
  expect($.types({})).toBe('object')
  // Test with false
  expect($.types(false)).toBe('boolean')
  // Test with function
  expect($.types(() => true)).toBe('function')
  // Test with immediately invoked function
  expect($.types((() => true)())).toBe('boolean')
  // Test with immediately invoked function in array
  expect($.types([(() => true)()])).toEqual(['boolean'])
  // Test with no parameter
  expect($.types()).toEqual([])
})
