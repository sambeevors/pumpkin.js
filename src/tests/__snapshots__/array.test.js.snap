// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`First Item(s) In Array 1`] = `
<li>
  Go shopping
</li>
`;

exports[`Last Item(s) In Array 1`] = `
<li>
  Raid fridge!
</li>
`;

exports[`Strip Items(s) From Array 1`] = `
Array [
  <li>
    Do laundry
  </li>,
  <li>
    Raid fridge!
  </li>,
]
`;
